lib/src/kbasesearchengine/main/NarrativeInfoDecorator.java

```diff
@@ -96,51 +105,64 @@ public SearchObjectsOutput searchObjects(final SearchObjectsInput params, final
     public GetObjectsOutput getObjects(final GetObjectsInput params, final String user)
             throws Exception {
         GetObjectsOutput getObjsOutput = searchInterface.getObjects(params, user);
+        final List<String> inputArr = new ArrayList<>();
+        if ("true".equals(System.getenv(removedGuidsEnv))){
+            getObjsOutput.getAdditionalProperties().put(removedGuids, inputArr);
+        }
         if (params.getPostProcessing() != null) {
             if (params.getPostProcessing().getAddNarrativeInfo() != null &&
                     params.getPostProcessing().getAddNarrativeInfo() == 1) {
                 getObjsOutput = getObjsOutput.withAccessGroupNarrativeInfo(addNarrativeInfo(
                         getObjsOutput.getObjects(),
-                        getObjsOutput.getAccessGroupNarrativeInfo()));
+                        getObjsOutput.getAccessGroupNarrativeInfo(),
+                        inputArr));
             }
         }
         return getObjsOutput;
     }
 
+    /**
+     * Adds narrative information for non deleted and valid narrative workspaces. Removes results otherwise and
+     * log list of removed guids. If env "KBASE_SEARCH_SHOW_REMOVED_GUIDS", is set to true, list of removed guids
+     * is added to additionalProperties.
+     */
     private Map<Long, Tuple5 <String, Long, Long, String, String>> addNarrativeInfo(
             final List<ObjectData> objects,
-            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo) {
+            final Map<Long, Tuple5 <String, Long, Long, String, String>> accessGroupNarrInfo,
+            final List<String> removedGuids) {
 
         final Map<Long, Tuple5 <String, Long, Long, String, String>> retVal = new HashMap<>();
 
         if (accessGroupNarrInfo != null) {
             retVal.putAll(accessGroupNarrInfo);
         }
-        final Set<Long> wsIdsSet = new HashSet<>();
-
-        for (final ObjectData objData: objects) {
+        final Set<String> userNames = new HashSet<>();
+        final Iterator<ObjectData> iter = objects.iterator();
+        while (iter.hasNext()) {
+            final ObjectData objData = iter.next();
             final GUID guid = new GUID(objData.getGuid());
             if (WorkspaceEventHandler.STORAGE_CODE.equals(guid.getStorageCode())) {
-                wsIdsSet.add((long) guid.getAccessGroupId());
-            }
-        }
-        final Set<String> userNames = new HashSet<>();
-        for (final long workspaceId: wsIdsSet) {
```

Comment:
Why is the 2 stage loop removed? That'll result in up to  9999 more calls to the narrative info provider than necessary.

Reply:
narrative info provider is a cache with size 50000 * 1000 (why is it so large??), so it should be fine. Otherwise, need to keep track of deleted ws ids, and then go through the again to remove all the objects in deleted ws.